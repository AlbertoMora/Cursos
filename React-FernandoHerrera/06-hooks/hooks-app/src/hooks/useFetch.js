import { useEffect, useState, useRef } from "react"

export const useFetch = (url) => {

    const [state, setState] = useState({data: null, loading: true, error: null});

    const isMounted = useRef(true);


    useEffect(() => {
        
        return () => {
            isMounted.current = false;
        }
    }, [])

    useEffect(() => {
        
        setState({
            ...state,
            loading: true
        });

        fetch(url)
            .then(resp => resp.json())
            .then(data => {
                //Previene un setState en un componente desmontando.
                if(isMounted.current){
                    setState({
                        loading: false,
                        error: null,
                        data
                    });
                }
            })
            .catch(err => {
                setState({
                    data: null,
                    loading: false,
                    error: 'Info could not be fetched'
                })
            });

    }, [url]);

    return state;

}
