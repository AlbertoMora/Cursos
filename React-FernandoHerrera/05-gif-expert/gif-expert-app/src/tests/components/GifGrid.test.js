import React from "react";
import { shallow } from "enzyme";

import GifGrid from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock("../../hooks/useFetchGifs");

describe("Test on GifGrid", () => {
    const category = "Goku";

    test("should match snapshot", () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true,
        });
        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
    });
    test("should show items when images are loaded", () => {
        const gifs = [
            {
                id: "ABC",
                url: "https://cnd.images.com/as.jpg",
                title: "Anything",
            },
            {
                id: "123",
                url: "https://cnd.images.com/as.jpg",
                title: "Anything",
            },
        ];
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false,
        });

        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
    });
});
