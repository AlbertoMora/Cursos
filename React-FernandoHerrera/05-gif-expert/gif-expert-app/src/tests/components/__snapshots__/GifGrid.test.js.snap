// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on GifGrid should match snapshot 1`] = `
<Fragment>
  <h3>
    Goku
  </h3>
  <p>
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
